src/cargo/util/toml_mut/manifest.rs

```diff
@@ -409,6 +411,49 @@ impl LocalManifest {
                 }
             }
         }
+
+        // continue only if the dep is fully removed
+        if status != DependencyStatus::None {
+            return;
+        }
+
+        // clean up profile sections
+        let profile_table_path = [
+            "profile".to_owned(),
+            "dev".to_owned(),
+            "package".to_owned(),
+            dep_key.to_owned(),
+        ];
+        if let Ok(item @ toml_edit::Item::Table(_)) = self.get_table_mut(&profile_table_path) {
+            *item = toml_edit::Item::None
+        }
+
+        // clean up patch sections
+        if let Some(toml_edit::Item::Table(patch_table)) = self.data.get_mut("patch") {
+            patch_table.set_implicit(true);
+            for (_, item) in patch_table.iter_mut() {
+                if let toml_edit::Item::Table(table) = item {
+                    table.set_implicit(true);
+                    for (key, item) in table.iter_mut() {
+                        if key.get() == dep_key {
+                            *item = toml_edit::Item::None;
+                        }
+                    }
+                }
+            }
+        }
+
+        // clean up replace section
+        if let Some(toml_edit::Item::Table(replace_table)) = self.data.get_mut("replace") {
+            replace_table.set_implicit(true);
+            for (key, item) in replace_table.iter_mut() {
+                if let Ok(spec) = PackageIdSpec::parse(key.get()) {
+                    if spec.name() == dep_key {
```

Comment:
Could the user be `replace`ing one of several versions of a dependency in the dependency tree?

Reply:
I assume so. This solution works regardless of whether there can be duplicates or not though, so I'm not quite sure what you're getting at.

Reply:
Say I have a CLI that has benchmarks.  I am directly depending on clap 4.0.  I also have a dev-dependency on criterion 0.4 which depends on clap 3.2.  My lockfile will have a clap 4.0.10 entry and a 3.2.18 entry.

Say I want to `replace` criterion's clap 3.2.18 entry with one from a git repo:
```toml
[replace]
"clap:3.2.18" = { git = "https://github.com/clap-rs/clap.git" }
```

If I then do `cargo rm clap`, I'll be removing a clap 4 version requirement but also a clap 3 replace entry.